neat_core.c

```diff
@@ -465,6 +495,139 @@ static void io_writable(neat_ctx *ctx, neat_flow *flow,
+// into NEAT error codes
+static neat_error_code sctp_to_neat_code(uint16_t sctp_code)
+{
+    neat_error_code outcode;
+
+    // TODO: this translation table is not very good,
+    // should probably improve on it, add more NEAT codes
+    switch (sctp_code) {
+    case SCTP_CAUSE_INVALID_STREAM:
+    case SCTP_CAUSE_UNRESOLVABLE_ADDR:
+    case SCTP_CAUSE_UNRECOG_CHUNK:
+    case SCTP_CAUSE_INVALID_PARAM:
+    case SCTP_CAUSE_UNRECOG_PARAM:
+    case SCTP_CAUSE_NO_USER_DATA:
+    case SCTP_CAUSE_MISSING_PARAM:
+    case SCTP_CAUSE_STALE_COOKIE:
+	outcode = NEAT_ERROR_BAD_ARGUMENT;
+	break;
+    case SCTP_CAUSE_OUT_OF_RESC:
+	outcode = NEAT_ERROR_IO;
+	break;
+    case SCTP_CAUSE_COOKIE_IN_SHUTDOWN:
+    case SCTP_CAUSE_PROTOCOL_VIOLATION:
+    case SCTP_CAUSE_RESTART_W_NEWADDR:
+	outcode = NEAT_ERROR_INTERNAL;
+	break;
+    case SCTP_CAUSE_USER_INITIATED_ABT:
+	outcode = NEAT_ERROR_REMOTE;
+	break;
+    default:
+	outcode = NEAT_ERROR_INTERNAL;
+    }
+
+    return outcode;
+}
+
+
+// Handle SCTP association change events
+// includes shutdown complete, etc.
+static void handle_sctp_assoc_change(neat_flow *flow, struct sctp_assoc_change *sac)
+{
+    neat_log(NEAT_LOG_DEBUG, "%s", __func__);
+
+    switch (sac->sac_state) {
+    case SCTP_SHUTDOWN_COMP:
+	neat_notify_close(flow);
+	break;
+    case SCTP_COMM_LOST:
+	// Draft specifies to return cause code, D1.2 doesn't - we
+	// follow D1.2
+	neat_notify_aborted(flow);
+	// Fallthrough:
+    case SCTP_COMM_UP: // Fallthrough:
+    case SCTP_RESTART:
+	// TODO: might want to "translate" the state codes to a NEAT code.
+	neat_notify_network_changed(flow, sac->sac_state);
+	break;
+    }
+}
+
+// Handle SCTP send failed event
+// One is generated per failed message
+// Only FreeBSD and USRSCTP support the new RFC6458 API,
+// hence the ifdefs
+#if defined(__FreeBSD__) || defined (USRSCTP_SUPPORT)
+static void handle_sctp_send_failed(neat_flow *flow, struct sctp_send_failed_event *ssfe)
+#else
+static void handle_sctp_send_failed(neat_flow *flow, struct sctp_send_failed *ssf)
+#endif
+{
+    uint32_t error, context;
+    uint8_t *unsent_msg;
+    neat_log(NEAT_LOG_DEBUG, "%s", __func__);
+
+#if defined(__FreeBSD__) || defined (USRSCTP_SUPPORT)
+    error = ssfe->ssfe_error;
+    unsent_msg = ssfe->ssfe_data;
+    context = ssfe->ssfe_info.snd_context;
+#else
+    error = ssf->ssf_error;
+    unsent_msg = ssf->ssf_data;
+    context = ssf->ssf_info.sinfo_context;
+#endif
+
+    // TODO: context no. needs to be implemneted in the write functions!
+
+    neat_notify_send_failure(flow, sctp_to_neat_code(error), context, unsent_msg);
+}
+
+// Handle notifications about SCTP events
+static void handle_sctp_event(neat_flow *flow, union sctp_notification *notfn)
+{
+    neat_log(NEAT_LOG_DEBUG, "%s", __func__);
+
+    switch (notfn->sn_header.sn_type) {
+    case SCTP_ASSOC_CHANGE:
+	handle_sctp_assoc_change(flow, &notfn->sn_assoc_change);
+	break;
+#if defined(__FreeBSD__) || defined (USRSCTP_SUPPORT)
```

Comment:
I would suggest to base the #ifdef on the existence of the constants, not on the platforms.